### PR TITLE
Rewrite `countDistinctIf` with `count_distinct_implementation`  configuration.

### PR DESCRIPTION
**Pull request by Eridanus117**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Rewrite `countDistinctIf` with `count_distinct_implementation`  configuration.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@Eridanus117 yes, let's introduce a setting to enable or disable rewriting.
See also `compatibility` (the setting should be disabled if compatibility is less than 22.10).

File: src/Core/Settings.h [hunk 1]
```diff
@@ -480,6 +480,7 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(Bool, optimize_move_functions_out_of_any, false, "Move functions out of aggregate functions 'any', 'anyLast'.", 0) \
     M(Bool, optimize_normalize_count_variants, true, "Rewrite aggregate functions that semantically equals to count() as count().", 0) \
     M(Bool, optimize_injective_functions_inside_uniq, true, "Delete injective functions of one argument inside uniq*() functions.", 0) \
+    M(Bool, optimize_countdistinctif_rewrite, false, "Rewrite countDistinctIf with count_distinct_implementation configuration", 0) \
```
Comment (kitaisreal) [inline src/Core/Settings.h:483]: Probably better to name setting like `optimize_rewrite_count_distinct_if`.

File: tests/queries/0_stateless/01259_combinator_distinct.sql [hunk 2]
```diff
@@ -12,3 +12,11 @@ EXPLAIN SYNTAX SELECT countIf(DISTINCT number % 10, number % 5 = 2) FROM numbers
...
 SELECT sumIf(DISTINCT number % 10, number % 5 = 2) FROM numbers(10000);
 EXPLAIN SYNTAX SELECT sumIf(DISTINCT number % 10, number % 5 = 2) FROM numbers(10000);
+
+SET optimize_countdistinctif_rewrite = FALSE;
+SELECT countDistinctIf(number % 10, number % 5 = 2) FROM numbers_mt(100000000);
+EXPLAIN SYNTAX SELECT countDistinctIf(number % 10, number % 5 = 2) FROM numbers_mt(100000000);
+
+SET optimize_countdistinctif_rewrite = TRUE;
+SELECT countDistinctIf(number % 10, number % 5 = 2) FROM numbers_mt(100000000);
+EXPLAIN SYNTAX SELECT countDistinctIf(number % 10, number % 5 = 2) FROM numbers_mt(100000000);
```
Comment (kitaisreal) [inline tests/queries/0_stateless/01259_combinator_distinct.sql:22]: Need to add new line.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Core/Settings.h (hunk 1 above)
Comment (ucasfl) [reply src/Core/Settings.h:483]: And I think it's better to enable it by default.

File: tests/queries/0_stateless/01259_combinator_distinct.sql (hunk 2 above)
Comment (kitaisreal) [reply tests/queries/0_stateless/01259_combinator_distinct.sql]: @ucasfl it is standard, and also a lot of text processing utilities expect for file to have new line at the end.

File: tests/queries/0_stateless/01259_combinator_distinct.sql [hunk 3]
```diff
@@ -12,3 +12,12 @@ EXPLAIN SYNTAX SELECT countIf(DISTINCT number % 10, number % 5 = 2) FROM numbers
 
 SELECT sumIf(DISTINCT number % 10, number % 5 = 2) FROM numbers(10000);
 EXPLAIN SYNTAX SELECT sumIf(DISTINCT number % 10, number % 5 = 2) FROM numbers(10000);
+
+SET optimize_rewrite_count_distinct_if = FALSE;
```
Comment (kitaisreal) [inline tests/queries/0_stateless/01259_combinator_distinct.sql:16]: Better to add new test, without modifying old tests. It is necessary for us to simplify tracking if some test becomes flacky.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

**Review by kitaisreal — APPROVED** [review]

**Comment by ucasfl** [comment]
@Eridanus117 Need fix conflict.

File: src/Core/SettingsChangesHistory.h [hunk 4]
```diff
@@ -90,6 +90,7 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
                {"query_plan_aggregation_in_order", 0, 1, "Enable some refactoring around query plan"},
                {"format_binary_max_string_size", 0, 1_GiB, "Prevent allocating large amount of memory"}}},
     {"22.11", {{"use_structure_from_insertion_table_in_table_functions", 0, 2, "Improve using structure from insertion table in table functions"}}},
+    {"22.10", {{"optimize_rewrite_count_distinct_if", false, true, "Rewrite countDistinctIf with count_distinct_implementation configuration"}}},
```
Comment (ucasfl) [inline src/Core/SettingsChangesHistory.h:93]: Here need to modify to the latest version.

**Comment by ucasfl** [comment]
LGTM. @kitaisreal Hi, what anything else need to do about this PR?

**Comment by alexey-milovidov** [comment]
@ucasfl the only remaining thing is to check if it works with Analyzer. If it works, let's merge.
Also, one minor concern that I had when reviewing - the setting is named "optimize" while it is not exactly an optimization...

**Comment by ucasfl** [comment]
Continued in #46051.

**Pull request closed without merging**